tests/test_generics.py

```diff
@@ -0,0 +1,284 @@
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class T2(Generic[_TValue, _TItem], T1[_TItem]):
+        value: _TValue
+        iterable: set[_TItem]
+
+    actual = get_fields_type_map(T2[Value2, int])
+    assert actual == {
+        "value": Value2,
+        "iterable": set[int],
+    }
+
+
+def test_get_fields_type_map_non_generic() -> None:
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class Value1:
+        v1: int
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class Value2(Value1):
+        v2: bool
+
+    actual = get_fields_type_map(Value2)
+    assert actual == {
+        "v1": int,
+        "v2": bool,
+    }
+
+
+def test_get_fields_type_map_generic_inheritance() -> None:
+    _T = TypeVar("_T")
+
+    @dataclasses.dataclass()
+    class NonGeneric:
+        v: bool
+
+    @dataclasses.dataclass()
+    class Value1(Generic[_T]):
+        v1: _T
+
+    @dataclasses.dataclass()
+    class Value2(Value1[int], NonGeneric):
+        v2: float
+
+    actual = get_fields_type_map(Value2)
+    assert actual == {
+        "v": bool,
+        "v1": int,
+        "v2": float,
+    }
+
+
+def test_get_fields_type_map_non_dataclass() -> None:
+    with pytest.raises(ValueError) as e:
+        get_fields_type_map(list[int])
+    assert e.value.args[0] == "<class 'list'> is not a dataclass"
+
+
+def test_get_fields_type_map_not_subscripted() -> None:
+    _T = TypeVar("_T")
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class Xxx(Generic[_T]):
+        xxx: _T
+
+    with pytest.raises(Exception) as e:
+        get_fields_type_map(Xxx)
+
+    assert e.value.args[0] == (
+        "Expected subscripted generic, but got unsubscripted "
+        "<class 'tests.test_generics.test_get_fields_type_map_not_subscripted.<locals>.Xxx'>"
+    )
+
+
+def test_get_fields_type_map_for_subscripted() -> None:
+    _T = TypeVar("_T")
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class Xxx(Generic[_T]):
+        xxx: _T
+
+    actual = get_fields_type_map(Xxx[str])
+    assert actual == {"xxx": str}
+
+
+def test_get_fields_class_map() -> None:
+    _T = TypeVar("_T")
+
+    @dataclasses.dataclass()
+    class Base1(Generic[_T]):
+        a: str
+        b: str
+        c: str
+
+    @dataclasses.dataclass()
+    class Base2(Base1[int]):
+        a: str
+        b: str
+        d: str
+        e: str
+
+    @dataclasses.dataclass()
+    class BaseG:
+        f: str
+        g: str
+
+    @dataclasses.dataclass()
+    class Base3(Base2, BaseG):
+        a: str
+        d: str
+        f: str
+        h: str
+
+    actual = get_fields_class_map(Base3)
+    assert actual == {
+        "a": Base3,
+        "b": Base2,
+        "c": Base1,
+        "d": Base3,
+        "e": Base2,
+        "f": Base3,
+        "g": BaseG,
+        "h": Base3,
+    }
+
+
+def test_get_class_type_var_map_with_inheritance() -> None:
```

Comment:
@slawwan pls consider the following test case:
```python
def test_get_class_type_var_map_with_repeated_inheritance() -> None:
    _T1 = TypeVar("_T1")

    @dataclasses.dataclass()
    class Aaa(Generic[_T1]):
        pass

    @dataclasses.dataclass()
    class Bbb(Generic[_T1], Aaa[float]):  # Aaa
        pass

    @dataclasses.dataclass()
    class Ccc(Generic[_T1], Bbb[_T1], Aaa[_T1]):  # Aaa with different type args
        pass

    actual = get_class_type_var_map(Ccc[bool])
    assert actual == {
        Aaa: {
            _T1: float,
        },
        Bbb: {
            _T1: bool,
        },
        Ccc: {
            _T1: bool,
        },
    }
```

Reply:
same:

```python

def test_dump_load_with_field_override2() -> None:
    _T1 = TypeVar("_T1")

    @dataclasses.dataclass()
    class T1(Generic[_T1]):
        val1: _T1

    @dataclasses.dataclass()
    class T2(Generic[_T1]):
        pass

    @dataclasses.dataclass()
    class T3(T2[_T1], T1[int]):
        pass

    actual = get_fields_type_map(T3[str])
    assert actual == {
        "val1": str,
    }
```

Reply:
about first test case. I thought about it and this case looks invalid, also typechecker like pyright reports an error about this inheritance

```
error: Base classes of Ccc are mutually incompatible
    Base class "Aaa[_T1@Ccc]" derives from "Aaa[_T1@Ccc]" which is incompatible with type "Aaa[float]" (reportGeneralTypeIssues)
```

Reply:
I have made some improvements https://github.com/anna-money/marshmallow-recipe/pull/163/commits/461e0d9d085793577f9e58257a7c79ff01da2822

Reply:
after https://github.com/anna-money/marshmallow-recipe/commit/461e0d9d085793577f9e58257a7c79ff01da2822 provided case2 also will raise an error because of incompatible base classes

```ValueError: Incompatible Base class <class 'tests.test_generics.test_dump_load_with_field_override2.<locals>.T1'> with generic args {~_T1: <class 'str'>} and {~_T1: <class 'int'>}
```